gridfs.go

```diff
@@ -0,0 +1,391 @@
+	}
+	defer f.Close()
+
+	return errors.WithStack(b.Put(ctx, name, f))
+}
+
+func (b *gridfsBucket) Download(ctx context.Context, name, path string) error {
+	reader, err := b.Reader(ctx, name)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err = os.MkdirAll(filepath.Dir(path), 0700); err != nil {
+		return errors.Wrapf(err, "problem creating enclosing directory for '%s'", path)
+	}
+
+	f, err := os.Create(path)
+	if err != nil {
+		return errors.Wrapf(err, "problem creating file '%s'", path)
+	}
+	_, err = io.Copy(f, reader)
+	if err != nil {
+		_ = f.Close()
+		return errors.Wrap(err, "problem copying data")
+	}
+
+	return errors.WithStack(f.Close())
+}
+
+func (b *gridfsBucket) Push(ctx context.Context, local, remote string) error {
+	localPaths, err := walkLocalTree(ctx, local)
+	if err != nil {
+		return errors.Wrap(err, "problem finding local paths")
+	}
+
+	for _, path := range localPaths {
+		target := filepath.Join(remote, path)
+		_ = b.Remove(ctx, target)
+		if err = b.Upload(ctx, target, filepath.Join(local, path)); err != nil {
+			return errors.Wrapf(err, "problem uploading '%s' to '%s'", path, target)
+		}
+	}
+
+	if b.opts.DeleteOnSync && !b.opts.DryRun {
+		return errors.Wrapf(os.RemoveAll(local), "problem removing '%s' after push", local)
+	}
+
+	return nil
+}
+
+func (b *gridfsBucket) Pull(ctx context.Context, local, remote string) error {
+	iter, err := b.List(ctx, remote)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	keys := []string{}
+
+	for iter.Next(ctx) {
+		item := iter.Item()
+		name := filepath.Join(local, item.Name()[len(remote)+1:])
+		keys = append(keys, item.Name())
+
+		if err = b.Download(ctx, item.Name(), name); err != nil {
+			return errors.WithStack(err)
+		}
+	}
+
+	if err = iter.Err(); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if b.opts.DeleteOnSync && !b.opts.DryRun {
+		return errors.Wrapf(b.RemoveMany(ctx, keys...), "problem removing '%s' after pull", remote)
+	}
+
+	return nil
+}
+
+func (b *gridfsBucket) Copy(ctx context.Context, opts CopyOptions) error {
+	from, err := b.Reader(ctx, opts.SourceKey)
+	if err != nil {
+		return errors.Wrap(err, "problem getting reader for source")
+	}
+
+	to, err := opts.DestinationBucket.Writer(ctx, opts.DestinationKey)
+	if err != nil {
+		return errors.Wrap(err, "problem getting writer for destination")
+	}
+
+	if _, err = io.Copy(to, from); err != nil {
+		return errors.Wrap(err, "problem copying data")
+	}
+
+	return errors.WithStack(to.Close())
+}
+
+func (b *gridfsBucket) Remove(ctx context.Context, key string) error {
+	grid, err := b.bucket(ctx)
+	if err != nil {
+		return errors.Wrap(err, "problem resolving bucket")
+	}
+
+	cursor, err := grid.Find(bson.M{"filename": key})
+	if err == mongo.ErrNoDocuments {
+		return nil
+	} else if err != nil {
+		return errors.Wrap(err, "problem finding file")
+	}
+
+	document := struct {
+		ID interface{} `bson:"_id"`
+	}{}
+
+	for cursor.Next(ctx) {
```

Comment:
Why is this written as a duplicate of the RemoveMany function?

Reply:
Remove handles missing keys differently.

Reply:
so filename is not unique here?

Reply:
It might not be.

Reply:
got it